### PR TITLE
Fix breakpoints inside .cshtml files.

### PR DESCRIPTION
**Pull request by KirillOsenkov**
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/984248

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Navigation/TextFileNavigationPoint.cs [hunk 1]
```diff
@@ -128,7 +128,9 @@ protected override async Task<Document> DoShow ()
 
 		protected void JumpToCurrentLocation (ITextView textView)
 		{
+#if !WINDOWS
```
Comment (avodovnik) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Navigation/TextFileNavigationPoint.cs]: This scares me a bit - as in, what do we do on Windows? Is it OK to just do _nothing_?

File: main/src/addins/MonoDevelop.TextEditor/MonoDevelop.TextEditor/ThemeToClassification.cs [hunk 2]
```diff
@@ -189,7 +189,9 @@ void UpdateEditorFormatMap (string appearanceCategory)
 			// Handling all this is very complicated and error prone, since this whole file should be removed at some point
 			// and replaced with new theming system better adjusted to ITextView
 			// See https://devdiv.visualstudio.com/DevDiv/_workitems/edit/980541 for more details
+#if !WINDOWS
 			(editorFormat as IEditorFormatMap2)?.ClearProperties ();
```
Comment (avodovnik) [inline main/src/addins/MonoDevelop.TextEditor/MonoDevelop.TextEditor/ThemeToClassification.cs]: What has the theme classification have to do with the breakpoints? Just curious here.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Navigation/TextFileNavigationPoint.cs (hunk 1 above)
Comment (KirillOsenkov) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Navigation/TextFileNavigationPoint.cs]: Yeah at this point I just want Windows to build. The bar is low.

File: main/src/addins/MonoDevelop.TextEditor/MonoDevelop.TextEditor/ThemeToClassification.cs (hunk 2 above)
Comment (KirillOsenkov) [reply main/src/addins/MonoDevelop.TextEditor/MonoDevelop.TextEditor/ThemeToClassification.cs]: It’s an omnibus PR. I’m sneaking stuff in, hoping no one will notice! Just fixing the build opportunistically.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/WindowCommands.cs [hunk 3]
```diff
@@ -253,9 +253,11 @@ internal class OpenWindowListHandler : CommandHandler
 		protected override void Update (CommandArrayInfo info)
 		{
 			foreach (Components.Window window in IdeApp.CommandService.TopLevelWindowStack) {
+#if !WINDOWS
 				//we don't want include hidden windows
 				if (!window.IsRealized || !window.IsVisible || Components.Mac.GtkMacInterop.IsGdkQuartzWindow (window))
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/WindowCommands.cs]: What we want is only the IsGtkQuartzWindow code to be under `#if MAC`

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/BreakpointManager.cs [hunk 4]
```diff
@@ -126,7 +129,8 @@ public void Dispose ()
 		public IEnumerable<BreakpointSpan> GetBreakpoints (ITextSnapshot snapshot)
 		{
 			foreach (var item in breakpoints.Values) {
-				yield return new BreakpointSpan (item.Breakpoint, item.TrackingSpan.GetSpan (snapshot));
+				if (item.TrackingSpan.TextBuffer == snapshot.TextBuffer)
```
Comment (avodovnik) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/BreakpointManager.cs]: I feel like we should be doing _something_ if this isn't true; maybe use the buffer graph to remap?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/WindowCommands.cs (hunk 3 above)
Comment (KirillOsenkov) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Commands/WindowCommands.cs]: Nah, as I said, at this point I pretty much only care for Windows to build. We have lost so much functionality that I don't expect it to actually run with any degree of reliability. I'm afraid that ship has sailed.

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/BreakpointManager.cs (hunk 4 above)
Comment (KirillOsenkov) [reply main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/BreakpointManager.cs]: This is called several times for each buffer. We throw away all the wrong buffers, and let the right buffer in. This is our highly scientific way to ensure the buffer matches ;)

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs [hunk 5]
```diff
@@ -53,14 +54,16 @@ public IEnumerable<ITagSpan<T>> GetTags (NormalizedSnapshotSpanCollection spans)
...
+					if (newTag != null && newTag.Span.Snapshot.TextBuffer == spans [0].Snapshot.TextBuffer)
+						if (newTag != null && spans.IntersectsWith (newTag.Span))
+							yield return newTag;
 				}
 			} else {
 				if (DebuggingService.CurrentFrameIndex == 0) {
 					var newTag = CreateTag ();
-					if (newTag != null && spans.IntersectsWith (newTag.Span))
-						yield return newTag;
+					if (newTag != null && newTag.Span.Snapshot.TextBuffer == spans [0].Snapshot.TextBuffer)
```
Comment (sandyarmstrong) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs]: Can we merge this `if` with the line below?

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs [hunk 6]
```diff
@@ -53,14 +54,16 @@ public IEnumerable<ITagSpan<T>> GetTags (NormalizedSnapshotSpanCollection spans)
 			if (isGreen) {
 				if (DebuggingService.CurrentFrameIndex > 0) {
 					var newTag = CreateTag ();
-					if (newTag != null && spans.IntersectsWith (newTag.Span))
-						yield return newTag;
+					if (newTag != null && newTag.Span.Snapshot.TextBuffer == spans [0].Snapshot.TextBuffer)
```
Comment (sandyarmstrong) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs]: Can we merge this `if` with the line below?

**Comment by slluis** [comment]
@monojenkins backport release-8.3

**Comment by KirillOsenkov** [comment]
Once we have a .dmg we will smoke test and report here.

**Comment by KirillOsenkov** [comment]
@monojenkins backport master

File: main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs [hunk 7]
```diff
@@ -53,13 +54,13 @@ public IEnumerable<ITagSpan<T>> GetTags (NormalizedSnapshotSpanCollection spans)
...
 					var newTag = CreateTag ();
-					if (newTag != null && spans.IntersectsWith (newTag.Span))
+					if (newTag != null && newTag.Span.Snapshot.TextBuffer == spans [0].Snapshot.TextBuffer && spans.IntersectsWith(newTag.Span))
 						yield return newTag;
 				}
 			} else {
 				if (DebuggingService.CurrentFrameIndex == 0) {
 					var newTag = CreateTag ();
-					if (newTag != null && spans.IntersectsWith (newTag.Span))
+					if (newTag != null && newTag.Span.Snapshot.TextBuffer == spans [0].Snapshot.TextBuffer && spans.IntersectsWith(newTag.Span))
```
Comment (sandyarmstrong) [inline main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs]: In retrospect, this whole thing can go inside something like `if (DebuggingService.CurrentFrameIndex == 0 || (isGreen && DebuggingService.CurrentFrameIndex > 0))`, but we should fix that later. I just didn't realize until now that it's identical code here.

Comment (KirillOsenkov) [reply main/src/addins/MonoDevelop.Debugger/MonoDevelop.Debugger.VSTextView/Tags/AbstractCurrentStatementTagger.cs]: Honestly if I'd started to refactor that code I would never stop... there needs to be a systemic sweep all over that code in 8.4

**Comment by sandyarmstrong** [comment]
@monojenkins backport master

**Pull request merged**